semi join query throws AssertionError

Example query:
```sql
SELECT COUNT(*)
FROM (
  SELECT DISTINCT dim2
  FROM druid.foo
  WHERE SUBSTRING(dim2, 1, 1) IN (
    SELECT SUBSTRING(dim1, 1, 1) FROM druid.foo WHERE dim1 <> ''
  ) AND __time >= '2000-01-01' AND __time < '2002-01-01'
)
```

When outer query contains AND filters like above, druid throws AssertionError when constructing Filter.

```
java.lang.AssertionError: AND(AND(>=($0, 2000-01-01 00:00:00), <($0, 2002-01-01 00:00:00)), OR(=(SUBSTRING($3, 1, 1), '1'), =(SUBSTRING($3, 1, 1), '2'), =(SUBSTRING($3, 1, 1), 'a'), =(SUBSTRING($3, 1, 1), 'd')))
	at org.apache.calcite.rel.core.Filter.<init>(Filter.java:74)
	at org.apache.calcite.rel.logical.LogicalFilter.<init>(LogicalFilter.java:71)
	at org.apache.calcite.rel.logical.LogicalFilter.copy(LogicalFilter.java:136)
	at org.apache.calcite.rel.logical.LogicalFilter.copy(LogicalFilter.java:48)
	at io.druid.sql.calcite.rel.DruidSemiJoin.getLeftRelWithFilter(DruidSemiJoin.java:345)
```

The reason is calcite requires filter condition to be flattened, while DruidSemiJoin#getLeftRelWithFilter may construct unflatten filter.

--- a/sql/src/main/java/io/druid/sql/calcite/rel/DruidSemiJoin.java
+++ b/sql/src/main/java/io/druid/sql/calcite/rel/DruidSemiJoin.java
@@ -44,6 +44,7 @@ import org.apache.calcite.rel.metadata.RelMetadataQuery;
 import org.apache.calcite.rel.type.RelDataType;
 import org.apache.calcite.rex.RexInputRef;
 import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexUtil;
 import org.apache.calcite.sql.fun.SqlStdOperatorTable;
 
 import javax.annotation.Nullable;
@@ -343,12 +344,15 @@ public class DruidSemiJoin extends DruidRel<DruidSemiJoin>
         newWhereFilter = whereFilter.copy(
             whereFilter.getTraitSet(),
             whereFilter.getInput(),
-            makeAnd(ImmutableList.of(whereFilter.getCondition(), makeOr(conditions)))
+            RexUtil.flatten(
+                getCluster().getRexBuilder(),
+                makeAnd(ImmutableList.of(whereFilter.getCondition(), makeOr(conditions)))
+            )
         );
       } else {
         newWhereFilter = LogicalFilter.create(
             leftPartialQuery.getScan(),
-            makeOr(conditions)
+            makeOr(conditions) // already in flattened form
         );
       }
 

--- a/sql/src/test/java/io/druid/sql/calcite/CalciteQueryTest.java
+++ b/sql/src/test/java/io/druid/sql/calcite/CalciteQueryTest.java
@@ -4650,7 +4650,7 @@ public class CalciteQueryTest extends CalciteTestBase
         + "  FROM druid.foo\n"
         + "  WHERE SUBSTRING(dim2, 1, 1) IN (\n"
         + "    SELECT SUBSTRING(dim1, 1, 1) FROM druid.foo WHERE dim1 <> ''\n"
-        + "  )\n"
+        + "  ) AND __time >= '2000-01-01' AND __time < '2002-01-01'\n"
         + ")",
         ImmutableList.of(
             GroupByQuery.builder()
@@ -4670,7 +4670,7 @@ public class CalciteQueryTest extends CalciteTestBase
                             new QueryDataSource(
                                 GroupByQuery.builder()
                                             .setDataSource(CalciteTests.DATASOURCE1)
-                                            .setInterval(QSS(Filtration.eternity()))
+                                            .setInterval(QSS(Intervals.of("2000-01-01/2002-01-01")))
                                             .setGranularity(Granularities.ALL)
                                             .setDimFilter(IN(
                                                 "dim2",

